src/util/invariant.h

```diff
@@ -51,14 +54,64 @@ Author: Martin Brain, martin.brain@diffblue.com
 ** CPROVER_INVARIANT_* macros.
 */
 
+/// A logic error, augmented with a distinguished field to hold a backtrace.
 class invariant_failedt: public std::logic_error
 {
+private:
+  /// Stores the backtrace set by `invariant_violated`.
+  std::string backtrace;
+
+  /// Exposes backtrace directly to the violated invariant dispatch method:
+  template<class T> friend
+  typename std::enable_if<std::is_base_of<invariant_failedt, T>::value>::type
+  invariant_violated(
+    const std::string &file,
+    const std::string &function,
+    const int line,
+    const T reason);
+
+  /// Temporary storage for `what`'s return value
+  mutable std::string what_str;
+
 public:
-  explicit invariant_failedt(const std::string& what) : logic_error(what) {}
-  explicit invariant_failedt(const char* what) : logic_error(what) {}
+  invariant_failedt(
+    const std::string &what):
+    logic_error(what)
+  {
+  }
+
+  /// Return a human-readable description of this error.
+  /// For this base class that's just the description string
+  /// supplied to the constructor; subclasses may override this to
+  /// provide a description in a different way, e.g. from a more
+  /// structured error description. `std::logic_error::what` is
+  /// overridden to use this and the backtrace to describe the
+  /// error.
+  virtual std::string pretty_print() const
+  {
+    return std::logic_error::what();
+  }
+
+  /// Describes this error using `backtrace` and `pretty_print`.
+  /// Cannot be overridden; override `pretty_print` instead.
+#ifdef _MSC_VER
```

Comment:
We should have a macro in `util` called `CBMC_NOEXCEPT` or similar which expands to `noexcept` on platforms which support it, and nothing on those which don't. We could do the same thing with `constexpr` too.
That way, this function declaration would become
```c++
const char *what() const CBMC_NOEXCEPT final
```
which is a bit cleaner.